### PR TITLE
Feature/tools 291 reduce required parameters

### PR DESCRIPTION
**Pull request by mpaddle**
### Ticket(s):
Link to ticket:
https://paddle.atlassian.net/browse/TOOLS-291

### Checklist
- [ ] I have added tests or left a comment on why they are not needed
- [ ] I deployed and tested the PR on development

### What has been done:
1. Ugly hacking about in the initial transaction queries:
    1. The non-flagged transactions are retrieved by different queries depending on which input parameters are given. We also try to speed up the query conditionally, based on whether the last four digits were given.
    1. The flagged transactions are also retrieved by different queries depending on the input parameters, but there is no need to attempt to speed it up: the transactions are always card transactions so the join to `spreedly_log` makes sense.
1. The input parameters are validated according to the new logic.

### How to test:

Note: on staging there is a lot of broken test data. Ledgers exist but not the matching orders. If this occurs, try to create an order with a more specific amount.

1. Deploy to staging.
1. Create a PayPal transaction with a reasonably unique amount.
1. Search for your transaction with a wide date range (e.g. a month) and just the amount.
1. Confirm that your transaction is included.
1. Create a card transaction with a reasonably unique amount.
1. Search for your transaction with a wide date range (e.g. a month) and just the amount.
1. Confirm that your transaction is included.
1. Search for your transaction with a narrow date range (e.g. a day) and just the last 4 digits.
1. Confirm that your transaction is included.
1. Search for your transaction with a wide date range (e.g. a day), the last 4 digits and amount.
1. Confirm that your transaction is included.

### CODE REVIEW

File: src/transaction_search/database/queries/monolith.py [hunk 1]
```diff
@@ -89,23 +111,40 @@ def _retrieve_base_nonflagged_transaction_data(
...
+        AND format(orders.payment_amount, 2) = format(:amount, 2)
+        """
+    elif trx_amount is not None:
+        parameters["amount"] = trx_amount
+        query_where = "format(orders.payment_amount, 2) = format(:amount, 2)"
+    elif last_four_digits is not None:
+        parameters["last_four_digits"] = last_four_digits
+        query_where = "last_four_digits = :last_four_digits"
+    else:
+        assert 0 == 1  # Not supposed to get here so prepare for failure.
```
Comment (vhemt) [inline src/transaction_search/database/queries/monolith.py]: could we raise a more relevant Exception?

File: src/transaction_search/database/queries/monolith.py [hunk 2]
```diff
@@ -12,61 +12,83 @@
...
+        opt_spreedly_join = "AND trx.type = 'SPREEDLY'"
+    elif trx_amount is not None:
+        parameters["amount"] = trx_amount
+        query_where = "format(trx.charged_amount, 2) = format(:amount, 2)"
+    elif last_four_digits is not None:
+        parameters["last_four_digits"] = last_four_digits
+        query_where = "card.last_four_digits = :last_four_digits"
+        opt_spreedly_join = "AND trx.type = 'SPREEDLY'"
+    else:
+        assert 0 == 1  # Not supposed to get here so prepare for failure.
```
Comment (vhemt) [inline src/transaction_search/database/queries/monolith.py]: could we raise a more relevant Exception?

File: src/transaction_search/database/queries/monolith.py (hunk 1 above)
Comment (mpaddle) [reply src/transaction_search/database/queries/monolith.py]: 👍 Done

File: src/transaction_search/database/queries/monolith.py (hunk 2 above)
Comment (mpaddle) [reply src/transaction_search/database/queries/monolith.py]: 👍 Done

**Pull request merged**